rfcs/0140-simple-package-paths.md

```diff
@@ -0,0 +1,242 @@
+author: Silvan Mosberger
+co-authors: Nixpkgs Architecture Team
+shepherd-team: (names, to be nominated and accepted by RFC steering committee)
+shepherd-leader: (name to be appointed by RFC steering committee)
+related-issues: https://github.com/NixOS/nixpkgs/pull/211832
+---
+
+# Summary
+[summary]: #summary
+
+Auto-generate trivial top-level attribute definitions in `pkgs/top-level/all-packages.nix` from a directory structure that matches the attribute name.
+This makes it much easier to contribute new packages packages, since there's no more guessing needed as to where the package should go, both in the ad-hoc directory categories and in `pkgs/top-level/all-packages.nix`.
+
+# Motivation
+[motivation]: #motivation
+
+- (Especially new) package contributors are having a hard time figuring out which files to add or edit. These are very common questions:
+  - Which directory should my package definition go in?
+  - What are all the categories and do they matter?
+  - What if the package has multiple matching categories?
+  - Why can't I build my package after adding the package file?
+  - Where in all-packages.nix should my package go?
+- Figuring out where an attribute is defined is a bit tricky:
+  - First one has to find the definition of it in all-packages.nix to see what file it refers to
+    - On GitHub this is even more problematic, as the `all-packages.nix` file is [too big to be displayed by GitHub](https://github.com/NixOS/nixpkgs/blob/nixos-22.05/pkgs/top-level/all-packages.nix)
+  - Then go to that file's definition, which takes quite some time for navigation (unless you have a plugin that can jump to it directly)
+    - It also slows down or even deadlocks editors due to the file size
+  - `nix edit -f . package-attr` works, though that's not yet stable (it relies on the `nix-command` feature being enabled) and doesn't work with packages that don't set `meta.position` correctly).
+- `all-packages.nix` frequently causes merge conflicts. It's a point of contention for all new packages
+
+# Detailed design
+[design]: #detailed-design
+
+This RFC establishes the standard of using `pkgs/unit/${shard}/${name}` "unit" directories for the definitions of the Nix packages `pkgs.${name}` in nixpkgs, where `shard = toLower (substring 0 2 name)`.
+All unit directories are automatically discovered and incorporated into the `pkgs` set using `pkgs.${name} = pkgs.callPackage pkgs/unit/${shard}/${name}/pkg-fun.nix { }`.
+
+The following requirements will be checked by CI.
+This standard must be followed for newly added packages that can satisfy these requirements.
+A treewide migration to this standard will be performed for existing packages that can satisfy these requirements.
+
+## Structure
+
+The `pkgs/unit` directory must only contain unit directories, and only in subdirectories of the form `${shard}/${name}`.
+Each unit directory must contain at least a `pkg-fun.nix` file, but may contain arbitrary other files and directories.
```

Comment:
IMO stick to ``default.nix``.

Reply:
Do note the discussion in https://github.com/nixpkgs-architecture/rfcs/blob/master/rfcs/0140-simple-package-paths.md#alternate-pkg-funnix-filename

Reply:
I don't like `pkg-fun.nix`. It's complicated because it's 2 words that are abbreviated. My first interpretation was "fun with packages" but i guess it stands for "function".

I think one englisch word would be better as file name. "package" is obvious.

I don't understand the explanation why that name is a bad idea.

Reply:
I agree with @davidak on the potential for confusion, and for something that will be so ubiquitous we should take the greatest care not to throw people off. But I can follow the argument to leave `pkg` or `package` for future extensions, and I also think one should be precise with the name to say what it means. And right now, packages are functions.

I suggest either a minimal change that makes it unambiguously "it's not fun":

```suggestion
Each unit directory must contain at least a `pkg-func.nix` file, but may contain arbitrary other files and directories.
```

or, and I find that idea better but it would be on you to specify it further, give the mechanism a name and call the file by that name, such as `callPackage` for the convention how the thing is used:

```suggestion
Each unit directory must contain at least a `callPackage.nix` file, but may contain arbitrary other files and directories.
```

Reply:
I like `callPackage.nix` more than `pkg.*-f.*n.*` variations

Reply:
I do like the suggestion of `pkg-func.nix`, but I can also see how a single word might be even better, not sure about `callPackage.nix`. I'll bring this up in the next [NAT meeting](https://github.com/nixpkgs-architecture/#meetings) on Monday.

Reply:
`pkg-func` is with with me

Reply:
`pkg-function.nix`, anyone?

Reply:
I also prefer `pkg-function.nix` over `pkg-fun.nix`, IMO `function` shouldn't be abbreviated as it can be hard coming to an agreement trying to choose from `fn//fun/func`.

I would also suggest `lambda.nix` and `pkg-lambda.nix` as alternatives because lambda is the type of the expression

```console
nix-repl> builtins.typeOf ({}: 42)
"lambda"
```

Reply:
Note also the discussion with @davidak [here](https://github.com/NixOS/rfcs/pull/140#discussion_r1098667630). I would like to reconsider `package.nix`.

I would also like to point out that these unit directories will arguably need versioning at some point (especially once they become a stable API for third-parties (not in scope for this RFC)). This will likely involve a `version.nix` file, where we could declare the version of the unit directory. With version 2, the `package.nix` file could have a different structure, deprecating the old one. Because of that I don't think we need to be very worried about using `package.nix`, so this is my current favorite.

Reply:
We just had the NAT meeting and reached consensus that `package.nix` is good, [notes here](https://discourse.nixos.org/t/nixpkgs-architecture-team-meeting-29/25439#pkg-funnix-vs-other-nameshttpsgithubcomnixosrfcspull140discussion_r1090915765-2) :slightly_smiling_face:

Reply:
I just pushed [a commit](https://github.com/NixOS/rfcs/pull/140/commits/967beb129b463a6096f1d3c74bc3d37343439904) to rename this to `package.nix`, I'll resolve this thread